components/de_amp/browser/test/de_amp_browsertest.cc

```diff
@@ -0,0 +1,274 @@
+    const std::string& canonical_page,
+    const std::string& body,
+    const net::test_server::HttpRequest& request) {
+  const auto port = request.GetURL().port();
+
+  auto http_response = std::make_unique<net::test_server::BasicHttpResponse>();
+  http_response->set_code(net::HTTP_OK);
+  http_response->set_content_type("text/html");
+  // ... rel="canonical" href="https://%s:%s%s" becomes
+  // href="https://<test server base url>:<test port>/<canonical html page>"
+  http_response->set_content(base::StringPrintf(
+      body.c_str(), base_url.c_str(), port.c_str(), canonical_page.c_str()));
+  return http_response;
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, SimpleDeAmp) {
+  TogglePref(true);
+  ASSERT_TRUE(https_server_->Start());
+
+  // Go to any page
+  const GURL simple = https_server_->GetURL(kTestHost, kTestSimpleNonAmpPage);
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), simple));
+  EXPECT_EQ(web_contents()->GetLastCommittedURL(), simple);
+
+  // Now go to an AMP page
+  https_server_.reset(new net::EmbeddedTestServer(
+      net::test_server::EmbeddedTestServer::TYPE_HTTPS));
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  const GURL landing_url = https_server_->GetURL(kTestHost, kTestCanonicalPage);
+  NavigateToURLAndWaitForRedirects(original_url, landing_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, NonHttpScheme) {
+  TogglePref(true);
+  const std::string nonHttpSchemeBody =
+      R"(
+        <html amp>\
+      <head>\
+      <link rel='canonical' 
+      href='brave://settings'>\
+      </head></html>
+      )";
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, nonHttpSchemeBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, CanonicalLinkSameAsAmpPage) {
+  TogglePref(true);
+  https_server_->RegisterRequestHandler(
+      base::BindRepeating(HandleRequest, kTestHost, kTestAmpPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, PrefOff) {
+  TogglePref(false);
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  // Doesn't get De-AMPed
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, RestorePage) {
```

Comment:
`RestoreTab` would be more clear here imo

Reply:
In fact we restore the whole window (by calling `OpenWindowWithRestoredTabs()`)